File: app/src/commonMain/kotlin/dev/audiocompanion/app/SegmentPlayback.kt
package dev.audiocompanion.app

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.concurrent.Volatile
import kotlin.math.min

/**
 * Platform PCM playback sink: 16 kHz mono signed 16-bit. [write] provides pacing — it must not
 * return until the device can accept more audio, which is what bounds decode-ahead memory.
 */
interface PcmAudioPlayer {
    fun start(sampleRateHz: Int)
    suspend fun write(pcm: ShortArray)
    fun setSpeed(speed: Float)
    fun stop()
}

data class PlaybackUiState(
    val segmentId: String? = null,
    val playing: Boolean = false,
    /** Media position within the stored audio (gaps excluded), not wall time. */
    val positionMs: Long = 0,
    val durationMs: Long = 0,
    val speed: Float = 1f,
)

/**
 * Segment playback with scrubbing and speed control (MVP requirement; ux plan Section 9).
 *
 * Decode is chunked — [BATCH_FRAMES] encoded frames at a time — never the whole segment in
 * memory. Seeking is frame-accurate because the firmware resets the Speex bitstream per frame,
 * so any frame is independently decodable.
 */
class SegmentPlaybackController(
    private val playerFactory: () -> PcmAudioPlayer,
    private val decoder: LiveFrameDecoder?,
    /** Encoded frame payloads of a segment, in order (from the durable frame log). */
    private val frameSource: (segmentId: String) -> List<ByteArray>,
    private val frameDurationMs: Long = 20,
    private val sampleRateHz: Int = 16_000,
) {
    private val _state = MutableStateFlow(PlaybackUiState())
    val state: StateFlow<PlaybackUiState> = _state.asStateFlow()

    private var scope: CoroutineScope? = null

    @Volatile
    private var seekFrameRequest: Int = -1

    /** Starts (or resumes) playback of [segmentId] from the current position. */
    fun play(segmentId: String) {
        val current = _state.value
        val startMs = if (current.segmentId == segmentId) current.positionMs else 0L
        startLoop(segmentId, startMs)
    }

    fun pause() {
        stopLoop()
        _state.value = _state.value.copy(playing = false)
    }

    fun stop() {
        stopLoop()
        _state.value = PlaybackUiState(speed = _state.value.speed)
    }

    fun seekTo(segmentId: String, positionMs: Long) {
        val frame = (positionMs / frameDurationMs).toInt().coerceAtLeast(0)
        if (_state.value.playing && _state.value.segmentId == segmentId) {
            seekFrameRequest = frame
        } else {
            _state.value = _state.value.copy(
                segmentId = segmentId,
                positionMs = frame * frameDurationMs,
            )
        }
    }

    fun cycleSpeed() {
        val next = when (_state.value.speed) {
            1f -> 1.5f
            1.5f -> 2f
            else -> 1f
        }
        _state.value = _state.value.copy(speed = next)
        currentPlayer?.setSpeed(next)
    }

    @Volatile
    private var currentPlayer: PcmAudioPlayer? = null

    private fun startLoop(segmentId: String, startMs: Long) {
        stopLoop()
        val frames = frameSource(segmentId)
        val durationMs = frames.size * frameDurationMs
        if (frames.isEmpty() || decoder == null) {
            _state.value = _state.value.copy(
                segmentId = segmentId,
                playing = false,
                durationMs = durationMs,
            )
            return
        }
        val startFrame = (startMs / frameDurationMs).toInt().coerceIn(0, frames.size - 1)
        _state.value = _state.value.copy(
            segmentId = segmentId,
            playing = true,
            positionMs = startFrame * frameDurationMs,
            durationMs = durationMs,
        )
        seekFrameRequest = -1
        val loopScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
        scope = loopScope
        loopScope.launch {
            val player = playerFactory()
            currentPlayer = player
            player.start(sampleRateHz)
            player.setSpeed(_state.value.speed)
            var completed = false
            try {
                var index = startFrame
                while (index < frames.size) {
                    val requested = seekFrameRequest
                    if (requested >= 0) {
                        seekFrameRequest = -1
                        index = requested.coerceIn(0, frames.size - 1)
                    }
                    val end = min(index + BATCH_FRAMES, frames.size)
                    val pcm = decoder.decode(frames.subList(index, end))
                    player.write(pcm)
                    index = end
                    _state.value = _state.value.copy(positionMs = index * frameDurationMs)
                }
                completed = true
            } finally {
                currentPlayer = null
                player.stop()
                _state.value = _state.value.copy(
                    playing = false,
                    positionMs = if (completed) 0 else _state.value.positionMs,
                )
            }
        }
    }

    private fun stopLoop() {
        scope?.cancel()
        scope = null
        currentPlayer?.stop()
        currentPlayer = null
    }

    companion object {
        /** 25 frames = 500 ms of audio decoded per batch. */
        const val BATCH_FRAMES = 25
    }
}

File: ios/PebbleAudioKit/Sources/LiveAudio/SegmentPlayback.swift
import Foundation

// Port of `app/.../SegmentPlayback.kt`.

/// Platform PCM playback sink: 16 kHz mono signed 16-bit. `write` provides pacing — it must not
/// return until the device can accept more audio, which is what bounds decode-ahead memory.
public protocol PcmAudioPlayer: Sendable {
    func start(sampleRateHz: Int)
    func write(_ pcm: [Int16]) async
    func setSpeed(_ speed: Float)
    func stop()
}

public struct PlaybackUiState: Sendable, Equatable {
    public var segmentId: String?
    public var playing: Bool
    /// Media position within the stored audio (gaps excluded), not wall time.
    public var positionMs: Int64
    public var durationMs: Int64
    public var speed: Float

    public init(
        segmentId: String? = nil,
        playing: Bool = false,
        positionMs: Int64 = 0,
        durationMs: Int64 = 0,
        speed: Float = 1
    ) {
        self.segmentId = segmentId
        self.playing = playing
        self.positionMs = positionMs
        self.durationMs = durationMs
        self.speed = speed
    }
}

/// Segment playback with scrubbing and speed control (MVP requirement; ux plan Section 9).
///
/// Decode is chunked — `batchFrames` encoded frames at a time — never the whole segment in
/// memory. Seeking is frame-accurate because the firmware resets the Speex bitstream per frame,
/// so any frame is independently decodable.
///
/// The Kotlin class guarded state with `StateFlow` + `@Volatile`; here a lock guards the same
/// fields and the playback loop runs in one Task.
public final class SegmentPlaybackController: @unchecked Sendable {
    /// 25 frames = 500 ms of audio decoded per batch.
    public static let batchFrames = 25

    private let playerFactory: @Sendable () -> PcmAudioPlayer
    private let decoder: LiveFrameDecoder?
    /// Encoded frame payloads of a segment, in order (from the durable frame log).
    private let frameSource: @Sendable (String) -> [[UInt8]]
    private let frameDurationMs: Int64
    private let sampleRateHz: Int

    private let lock = NSLock()
    private var _state = PlaybackUiState()
    private var stateContinuations: [UUID: AsyncStream<PlaybackUiState>.Continuation] = [:]
    private var loopTask: Task<Void, Never>?
    private var seekFrameRequest: Int = -1
    private var currentPlayer: PcmAudioPlayer?

    public init(
        playerFactory: @escaping @Sendable () -> PcmAudioPlayer,
        decoder: LiveFrameDecoder?,
        frameSource: @escaping @Sendable (String) -> [[UInt8]],
        frameDurationMs: Int64 = 20,
        sampleRateHz: Int = 16_000
    ) {
        self.playerFactory = playerFactory
        self.decoder = decoder
        self.frameSource = frameSource
        self.frameDurationMs = frameDurationMs
        self.sampleRateHz = sampleRateHz
    }

    /// Current state (the Kotlin `StateFlow.value`).
    public var state: PlaybackUiState { lock.withLock { _state } }

    /// Live state updates for the UI; each subscriber immediately receives the current value.
    public func stateUpdates() -> AsyncStream<PlaybackUiState> {
        AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            let id = UUID()
            lock.withLock {
                stateContinuations[id] = continuation
                continuation.yield(_state)
            }
            continuation.onTermination = { [weak self] _ in
                guard let self else { return }
                self.lock.withLock { _ = self.stateContinuations.removeValue(forKey: id) }
            }
        }
    }

    private func updateState(_ transform: (inout PlaybackUiState) -> Void) {
        let (snapshot, continuations): (PlaybackUiState, [AsyncStream<PlaybackUiState>.Continuation])
        lock.lock()
        transform(&_state)
        snapshot = _state
        continuations = Array(stateContinuations.values)
        lock.unlock()
        for continuation in continuations { continuation.yield(snapshot) }
    }

    /// Starts (or resumes) playback of `segmentId` from the current position.
    public func play(_ segmentId: String) {
        let current = state
        let startMs = current.segmentId == segmentId ? current.positionMs : 0
        startLoop(segmentId, startMs: startMs)
    }

    public func pause() {
        stopLoop()
        updateState { $0.playing = false }
    }

    public func stop() {
        stopLoop()
        updateState { $0 = PlaybackUiState(speed: $0.speed) }
    }

    public func seekTo(_ segmentId: String, positionMs: Int64) {
        let frame = max(Int(positionMs / frameDurationMs), 0)
        let current = state
        if current.playing && current.segmentId == segmentId {
            lock.withLock { seekFrameRequest = frame }
        } else {
            updateState {
                $0.segmentId = segmentId
                $0.positionMs = Int64(frame) * frameDurationMs
            }
        }
    }

    public func cycleSpeed() {
        let next: Float
        switch state.speed {
        case 1: next = 1.5
        case 1.5: next = 2
        default: next = 1
        }
        updateState { $0.speed = next }
        lock.withLock { currentPlayer }?.setSpeed(next)
    }

    private func startLoop(_ segmentId: String, startMs: Int64) {
        stopLoop()
        let frames = frameSource(segmentId)
        let durationMs = Int64(frames.count) * frameDurationMs
        guard !frames.isEmpty, let decoder else {
            updateState {
                $0.segmentId = segmentId
                $0.playing = false
                $0.durationMs = durationMs
            }
            return
        }
        let startFrame = min(max(Int(startMs / frameDurationMs), 0), frames.count - 1)
        updateState {
            $0.segmentId = segmentId
            $0.playing = true
            $0.positionMs = Int64(startFrame) * frameDurationMs
            $0.durationMs = durationMs
        }
        lock.withLock { seekFrameRequest = -1 }
        loopTask = Task { [weak self] in
            guard let self else { return }
            let player = self.playerFactory()
            self.lock.withLock { self.currentPlayer = player }
            player.start(sampleRateHz: self.sampleRateHz)
            player.setSpeed(self.state.speed)
            var completed = false
            defer {
                self.lock.withLock { self.currentPlayer = nil }
                player.stop()
                self.updateState {
                    $0.playing = false
                    if completed { $0.positionMs = 0 }
                }
            }
            var currentFrames = frames
            var index = startFrame
            while index < currentFrames.count {
                if Task.isCancelled { return }
                let requested = self.lock.withLock { () -> Int in
                    let value = self.seekFrameRequest
                    if value >= 0 { self.seekFrameRequest = -1 }
                    return value
                }
                if requested >= 0 {
                    index = min(max(requested, 0), currentFrames.count - 1)
                }
                let end = min(index + Self.batchFrames, currentFrames.count)
                let pcm = await decoder.decode(Array(currentFrames[index..<end]))
                if Task.isCancelled { return }
                await player.write(pcm)
                if Task.isCancelled { return }
                index = end
                self.updateState { $0.positionMs = Int64(index) * self.frameDurationMs }
                if index >= currentFrames.count {
                    // An open (still recording) segment may have stored more audio while
                    // we played: pick it up so replay can follow the live tail.
                    let refreshed = self.frameSource(segmentId)
                    if refreshed.count > currentFrames.count {
                        currentFrames = refreshed
                        let newDuration = Int64(currentFrames.count) * self.frameDurationMs
                        self.updateState { $0.durationMs = newDuration }
                    }
                }
            }
            completed = true
        }
    }

    private func stopLoop() {
        loopTask?.cancel()
        loopTask = nil
        let player = lock.withLock { () -> PcmAudioPlayer? in
            let value = currentPlayer
            currentPlayer = nil
            return value
        }
        player?.stop()
    }
}
